### PR TITLE
Add support for vecX<f16> and update vertex attributes 

### LINKED ISSUE #514: Add support for vec<f16> (aka vecXh)

### LINKED ISSUE #625: Add missing vertex formats
Implement [newly added](https://github.com/gpuweb/gpuweb/pull/4951) vertex formats

### PR DESCRIPTION
**Pull request by reczkok**
closes #514 
closes #625 
depends on #422 

TODO:
- [x] Implement new vertex formats
- [ ] Update data cheat sheet

### CODE REVIEW

File: packages/typegpu/src/shared/vertexFormat.ts [hunk 1]
```diff
@@ -96,25 +114,31 @@ type I32CompatibleFormats =
...
   | TgpuVertexAttrib<'float32x3'>
   | TgpuVertexAttrib<'float32x4'>
-  | TgpuVertexAttrib<'unorm10-10-10-2'>;
+  | TgpuVertexAttrib<'unorm10-10-10-2'>
+  | TgpuVertexAttrib<'unorm8x4-bgra'>;
 
 /**
  * All vertex attribute formats that can be interpreted as
- * an single or multi component f16 in a shader. (same as f32)
+ * an single or multi component f16 in a shader. (same as f32 on the shader side)
```
Comment (mhawryluk) [inline packages/typegpu/src/shared/vertexFormat.ts]:
```suggestion
 * a single or multi component f16 in a shader. (same as f32 on the shader side)
```

File: packages/typegpu/tests/vector.test.ts [hunk 2]
```diff
@@ -185,3 +186,71 @@ describe('vec4f', () => {
...
+
+  it('should encode a 2d vector', () => {
+    const vec = d.vec2h(1, 2050);
+    const buffer = new ArrayBuffer(sizeOf(d.vec2h));
+
+    writeData(new BufferWriter(buffer), d.vec2h, vec);
+    expect(readData(new BufferReader(buffer), d.vec2h)).toEqual(vec);
+  });
+
+  it('should change unrepresenable values to the closest representable value', () => {
```
Comment (mhawryluk) [inline packages/typegpu/tests/vector.test.ts]:
```suggestion
  it('should change unrepresentable values to the closest representable', () => {
```

File: packages/typegpu/tests/vector.test.ts [hunk 3]
```diff
@@ -185,3 +186,71 @@ describe('vec4f', () => {
...
+    // @ts-expect-error
+    acceptsVec2hSchema(d.vec2u);
+    // @ts-expect-error
+    acceptsVec2hSchema(d.vec2i);
+    // @ts-expect-error
+    acceptsVec2hSchema(d.vec2f);
+    // @ts-expect-error
+    acceptsVec2hSchema(d.vec3f);
+    // @ts-expect-error
+    acceptsVec2hSchema(d.vec4f);
```
Comment (mhawryluk) [inline packages/typegpu/tests/vector.test.ts:241]:
```suggestion
    acceptsVec2hSchema(d.vec3h);
    // @ts-expect-error
    acceptsVec2hSchema(d.vec4h);
```

File: packages/typegpu/tests/vertexLayout.test.ts [hunk 4]
```diff
@@ -301,4 +326,102 @@ describe('connectAttributesToShader', () => {
...
+    };
+
+    const layout = tgpu.vertexLayout((n) =>
+      d.looseArrayOf(
+        d.looseStruct({
+          alpha: d.f16, // 2 bytes
+          beta: d.float16x2, // 4 bytes
+          gamma: d.u32, // 4 bytes
+          delta: d.float16, // 2 bytes - ironically, this is interpreted as an f32
+          epsilon: d.float16, // 2bytes this is an f16
```
Comment (mhawryluk) [inline packages/typegpu/tests/vertexLayout.test.ts]:
```suggestion
          epsilon: d.float16, // 2 bytes - this is an f16
```

Comment (mhawryluk) [reply packages/typegpu/tests/vertexLayout.test.ts]: isn't this also f32 like delta?

**Review by mhawryluk — APPROVED** [review]
great job! 🔥

File: packages/typegpu/tests/vertexLayout.test.ts (hunk 4 above)
Comment (reczkok) [reply packages/typegpu/tests/vertexLayout.test.ts]: sure is 👍 this test is from before the removal of `h-attributes` - will fix

**Pull request merged**